### PR TITLE
Configurable check

### PR DESCRIPTION
**Pull request by becoded**
Allow to configure the check action.
Added flags:

      --disallowed_types  list of disallowed types
      --allowed_licenses list of allowed license names

Keeping the default behaviour of just checking the forbidden licenses.

### CODE REVIEW

File: check.go [hunk 1]
```diff
@@ -70,3 +111,228 @@ func checkMain(_ *cobra.Command, args []string) error {
...
+	for _, allowed := range allowedLicenseNames {
+		if allowed == licenseName {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getSupportedLicenseNames() []string {
```
Comment (Bobgy) [inline check.go:186]: Curious, why do we need this list?

Comment (becoded) [reply check.go:186]: I didn't find a method in github.com/google/licenseclassifier to get all defined license names.
It is used here to validate the input of flag `allowed_license_names`.

Comment (Bobgy) [reply check.go:186]: Is there a strong need to do this validation? If there is a typo, the check command fails and reports the correct license name in error message.

An extraneous invalid license name seems not much harm either.

**Comment by Bobgy** [comment]
Thank you for the contribution!

Let me start with high level questions.

Can you clarify the use case for allowed licenses? That seems too flexible as interface, it's probably only useful when very limited license types are allowed.

Would it be simpler asking someone with such flexible use case to directly parse the license CSV file and write custom validation?

**Comment by becoded** [comment]
Currently, only the list of forbidden licenses, defined in `google/licenseclassifier`, is validated.
And there is no way to change that, so it doesn't leave any room for flexibility in controlling that.
So for that reason, I added the other flags (`--exclude_forbidden`, `--exclude_notice`, `--exclude_permissive`, `--exclude_reciprocal`, `--exclude_restricted`, `--exclude_unencumbered`).
However, these sets are fixed.
I have a case where a subset is allowed and for others, we need special approval.
By allowing to create your own list of allowed licenses `--allowed_license_names` we are able to verify exactly what we need.

Personally I don't think it would be simpler to parse the report.
A part of this tool is written to check licenses, it makes sense too me that you can also control what is checked.

**Comment by Bobgy** [comment]
To explain why I'm asking, this tool was built primarily for Google's own use-cases for open source go projects, so I am seeking a balance between maintenance cost vs use-case coverage. I'm asking to learn more about why this feature is useful to you.

I think your use-case convinces me it may be useful to more people. Thank you for explaining that!

**Comment by Bobgy** [comment]
Regarding the interface, do the following changes make it slightly simpler and more consistent?

* Make exclude flags one flag that takes multiple values: `--disallowed_categories`, the description points to https://github.com/google/licenseclassifier/blob/e6a9bb99b5a6f71d5a34336b8245e305f5430f99/license_type.go#L180 for license categories.
* Reduce number of words on allowed_license_names: `--allowed_licenses`.

Also, I think we should add validation and description that, the two flags cannot be used at the same time.

**Comment by becoded** [comment]
I changed the interface. I did name it `types` instead of `categories` as that is what is used in multiple places in this repo and in `google/licenseclassifier`

**Comment by Bobgy** [comment]
nit: update PR description?

**Review by Bobgy — CHANGES_REQUESTED** [review]

File: check.go [hunk 2]
```diff
@@ -58,8 +86,13 @@ func checkMain(_ *cobra.Command, args []string) error {
...
 		}
 
-		if licenseType == licenses.Forbidden {
-			fmt.Fprintf(os.Stderr, "Forbidden license type %s for library %v\n", licenseName, lib)
+		if hasLicenseNames && !isAllowedLicenseName(licenseName, allowedLicenseNames) {
+			fmt.Fprintf(os.Stderr, "Not allowed license %s found for library %v\n", licenseName, lib)
+			found = true
+		}
+
+		if hasLicenseType && isDisallowedLicenseType(licenseType, disallowedLicenseTypes) {
```
Comment (Bobgy) [inline check.go:94]: See previous unmerged PR for a reference of how to handle unknown types: https://github.com/google/go-licenses/pull/44

File: README.md [hunk 3]
```diff
@@ -192,18 +194,44 @@ Report usage (using custom template file):
...
 go-licenses save <package> [package...] --save_path=<save_path>
 ```
 
+### Check
+
 Checking for forbidden licenses usage:
 
 ```shell
-go-licenses check <package> [package...]
+go-licenses check <package> [package...]
```
Comment (Bobgy) [inline README.md:210]: Also comment default behavior is disallow forbidden and unknown licenses?

**Comment by hansinator** [comment]
hey thank you for this PR and hopefully this gets merged soon, b/c we need to maintain a fork of license checker and it causes trouble with replace directives and deprecated go get / go install behaviour. basically soft-forks of "go install"-ed things are no longer a supported use-case since a few go versions so I'd really appreciate it if this goes upstream

File: README.md [hunk 4]
```diff
@@ -191,17 +193,42 @@ Report usage (using custom template file):
...
+Allow only specific license names:
+
+```shell
+go-licenses check <package> [package...] --allowed_license_names=<comma separated license names> 
+```
+
+* See supported license names: [github.com/google/licenseclassifier](https://github.com/google/licenseclassifier/blob/e6a9bb99b5a6f71d5a34336b8245e305f5430f99/license_type.go#L28)  
+
 
 Typically, specify the Go package that builds your Go binary.
```
[the remark this answers is not in the document]
Comment (becoded) [reply README.md]: Added it to the top of the usage block as it is helpful for all comments.

File: check.go [hunk 5]
```diff
@@ -16,29 +16,57 @@ package main
...
 		RunE:  checkMain,
 	}
+
+	allowedLicenses []string
+	disallowedTypes []string
 )
 
 func init() {
+	checkCmd.Flags().StringSliceVar(&allowedLicenses, "allowed_licenses", []string{}, "list of allowed license names, can't be used in combination with disallowed_types")
+	checkCmd.Flags().StringSliceVar(&disallowedTypes, "disallowed_types", []string{}, "list of disallowed license types, can't be used in combination with allowed_licenses")
```
[the remark this answers is not in the document]
Comment (becoded) [reply check.go]: Done

File: check.go [hunk 6]
```diff
@@ -70,3 +103,66 @@ func checkMain(_ *cobra.Command, args []string) error {
...
+		case "permissive":
+			excludedLicenseTypes = append(excludedLicenseTypes, licenses.Permissive)
+		case "reciprocal":
+			excludedLicenseTypes = append(excludedLicenseTypes, licenses.Reciprocal)
+		case "restricted":
+			excludedLicenseTypes = append(excludedLicenseTypes, licenses.Restricted)
+		case "unencumbered":
+			excludedLicenseTypes = append(excludedLicenseTypes, licenses.Unencumbered)
+		default:
+			fmt.Fprintf(os.Stderr, "Unknown license type '%s' provided", v)
```
[the remark this answers is not in the document]
Comment (becoded) [reply check.go]: Done

File: README.md (hunk 3 above)
Comment (becoded) [reply README.md:221]: Added it to the title

**Review by Bobgy — APPROVED** [review]
Awesome 👍!
Thank you for the great contribution!

**Pull request merged**